### DISCUSSION TITLE
Problem reading named pipes / FIFO files

### DISCUSSION THREAD

**Discussion by raddirad** [ipxe/ipxe#693] [General]
Hi,

I am poking around with iPXE and today I tried to make it read our automatically created named pipe files

```
# ls -l | grep 01-52-d4-7f-b9-82-07 
prw-r--r-- 1 root      root                 0 Jul 27 11:04 01-52-d4-7f-b9-82-07
```
```
# file 01-52-d4-7f-b9-82-07 
01-52-d4-7f-b9-82-07: fifo (named pipe)
```

Which when read has the following content
```
# cat 01-52-d4-7f-b9-82-07 
#!ipxe

kernel ../install-x64 initrd=miniroot-x64 video=vesa:ywrap,mtrr vga=791 quiet splash --no-log console=tty1 console=ttyS0 pckey=PCKEY hn=HOSTNAME dn=DOMAIN product=PRODUCT macaddress=MACADDRESS service=https://IPADDRESS:4447/rpc mem=7G
initrd ../miniroot-x64
boot
```

This named pipe is also server by the TFTP. On the client however I can se the named pipe is being read but it stalls at 5% of the transfer and thenn makes a fallback to the menu after 17 seconds. The named pipe is presend for 5 seconds after being read, which is currently a workaround when uding grub2 for UEFI netboot as the pipe is read more than once. This behaviour is also seen here.
```
Jul 27 11:08:10 SERVER in.tftpd[12745]: RRQ from 192.168.16.201 filename linux/pxelinux.cfg/01-52-d4-7f-b9-82-07
Jul 27 11:08:11 SERVER in.tftpd[12746]: RRQ from 192.168.16.201 filename linux/pxelinux.cfg/01-52-d4-7f-b9-82-07
Jul 27 11:08:27 SERVER in.tftpd[12751]: RRQ from 192.168.16.201 filename linux/pxelinux.cfg/pci-8086100e.ipxe
Jul 27 11:08:27 SERVER in.tftpd[12751]: sending NAK (1, File not found) to 192.168.16.201
Jul 27 11:08:27 SERVER in.tftpd[12752]: RRQ from 192.168.16.201 filename linux/pxelinux.cfg/chip-82540em.ipxe
Jul 27 11:08:27 SERVER in.tftpd[12752]: sending NAK (1, File not found) to 192.168.16.201
Jul 27 11:08:27 SERVER in.tftpd[12753]: RRQ from 192.168.16.201 filename linux/pxelinux.cfg/menu.ipxe
```

Is there any special config needed for the support of named pipes or FIFO files?
When the content is placed in a normal text file the UEFI netboot works fine

Regards
Mathias

**Comment by mcb30** [comment]
This is a TFTP issue, not an iPXE issue. TFTP is intrinsically a random-access protocol: blocks may be fetched more than once (eg in the event of lost packets). It is therefore not viable to have a TFTP server serving up something like a named pipe that supports only sequential access.

**Reply by raddirad** [reply]
We patched a named pipe / FIFO support into fttpd-hpa
It works fine with pxelinux and grub2. Therefore I'd expect to work with iPXE also.

**Reply by mcb30** [reply]
You have a bug in the patch to your TFTP server.

**Reply by raddirad** [reply]
This it the patch we use in tftpd-hpa
```
---
 tftpd/tftpd.c | 6 ++++++
 tftpd/tftpd.h | 1 +
 2 files changed, 7 insertions(+)

diff --git a/tftpd/tftpd.c b/tftpd/tftpd.c
index 364e7d2..522f053 100644
--- a/tftpd/tftpd.c
+++ b/tftpd/tftpd.c
@@ -1439,6 +1439,8 @@ static int validate_access(char *filename, int mode,
     const char **dirp;
     char stdio_mode[3];
 
+    int is_fifo = 0;
+
     tsize_ok = 0;
     *errmsg = NULL;
 
@@ -1509,6 +1511,10 @@ static int validate_access(char *filename, int mode,
             return (EACCESS);
         }
         tsize = stbuf.st_size;
+
+        if ( (is_fifo = S_ISFIFO(stbuf.st_mode)) ) {
+            tsize = FIFO_MAX_SIZE;
+        }
         /* We don't know the tsize if conversion is needed */
         tsize_ok = !pf->f_convert;
     } else {
diff --git a/tftpd/tftpd.h b/tftpd/tftpd.h
index e1d8bf0..07e7a7a 100644
--- a/tftpd/tftpd.h
+++ b/tftpd/tftpd.h
@@ -16,6 +16,7 @@
 
 #ifndef TFTPD_TFTPD_H
 #define TFTPD_TFTPD_H
+#define FIFO_MAX_SIZE 16384 /* Maximum bytes to read from a named pipe */
 
 void set_signal(int, void (*)(int), int);
 void *tfmalloc(size_t);
-- 
```

**Reply by mcb30** [reply]
Yeah, that patch is very very broken. You will be telling the client the incorrect filesize, terminating the transfer early, and failing to handle any retransmissions.

**Reply by raddirad** [reply]
Are you aware of a better patch/implementation?

Help here is very much appreciated :)

**Reply by mcb30** [reply]
Fundamentally it's not a sensible thing to try doing, since FIFOs by definition don't support `seek()`, and a TFTP server needs to be able to perform random accesses to the files that it's serving.  You could plausibly hack around it by having the TFTP server read the whole FIFO content into an in-memory buffer (or temporary file) and then serve block requests from that, but it would be a substantial amount of work.

The real question to ask is: why are you trying to do this in the first place?  There are many other mechanisms that can be used to dynamically generate iPXE scripts.  The most 'normal' way of doing it would be to have a web server running some server-side code (PHP, Python, Perl, Node.js, whatever language you prefer) that just generates and returns the script that you want.

**Reply by raddirad** [reply]
We are using it to deploy Windows and Linux operating system onto machines (https://opsi.org)
After the initial PXE/UEFI Netboot the TFTP bootfile should not be present to prevent another installation and let the machine finish the job. Therefore we are using named pipes, as they disappear after reading them once.

**Reply by mcb30** [reply]
That's not really a great design choice.  You have basically zero error handling - you have nothing checking that the deployment has actually run through to completion, and nothing preventing the named pipe from sitting around indefinitely if the client never tries to boot from it.

You have two sensible options, depending on how unattended you want to be.

Option 1: use the [`prompt`](https://ipxe.org/cmd/prompt) command in your script so that the user has to press a specific key to perform a reinstallation, and if the key is not pressed then the machine will just boot normally

Option 2: use something that keeps track of the deployed state, and will generate boot files as applicable based on centralised instructions.  I have never come across `opsi` before, so no idea if it supports this.

**Reply by raddirad** [reply]
Option 1: is not really doable for our customers and users. If they want to deploy a lot of machines at the same time no one would want to walk around and press buttons
Option 2: So what we have is an interface where users select the machines they want to deploy, set a OS product to setup and then the named pipe is created and vanishes after being read by the tftp. This was intended to let the installation perform completely unattended. I think we could rewrite some stuff to make "bootfiles" to boot  from local disc and boot from tftp/https when seomthing is set ti install
